internal/datacoord/services.go

```diff
@@ -1070,3 +1071,56 @@ func getDiff(base, remove []int64) []int64 {
 	}
 	return diff
 }
+
+// AcquireSegmentLock acquire the reference lock of the segments.
+func (s *Server) AcquireSegmentLock(ctx context.Context, req *datapb.AcquireSegmentLockRequest) (*commonpb.Status, error) {
+	resp := &commonpb.Status{
+		ErrorCode: commonpb.ErrorCode_UnexpectedError,
+	}
+
+	if s.isClosed() {
+		log.Warn("failed to acquire segments reference lock for closed server")
+		resp.Reason = msgDataCoordIsUnhealthy(Params.DataCoordCfg.GetNodeID())
+		return resp, nil
+	}
+
+	err := s.segRefer.AddSegmentsLock(req.SegmentIDs, req.NodeID)
+	if err != nil {
+		log.Warn("Add reference lock on segments failed", zap.Int64s("segIDs", req.SegmentIDs), zap.Error(err))
+		resp.Reason = err.Error()
+		return resp, nil
+	}
+	has := s.meta.HasSegments(req.SegmentIDs)
```

Comment:
Maybe I missed something...why not check this before `AddSegmentsLock()`?

Reply:
Because compaction may happen between `HasSegments` and `AddSegmentsLock`.

Reply:
Sure. I meant how about doing another `HasSegments` check before adding locks? If that check fails we don't need to  add any locks in the first place and save the trouble of add & releasing the lock.

Reply:
That's great, I will check it before acquire locks.